### PR TITLE
[GlobalIsel] Combine G_VSCALE

### PR DESCRIPTION
**Pull request by tschuett**
We need them for scalable address calculation and
legal scalable addressing modes.

### CODE REVIEW

File: llvm/include/llvm/CodeGen/GlobalISel/GenericMachineInstrs.h [hunk 1]
```diff
@@ -856,6 +858,43 @@ class GTrunc : public GCastOp {
...
+/// Represents an integer multiplication.
+class GMul : public GIntBinOp {
+public:
+  static bool classof(const MachineInstr *MI) {
+    return MI->getOpcode() == TargetOpcode::G_MUL;
+  };
+};
+
+/// Represents a shift left.
+class GSHL : public GenericMachineInstr {
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/GlobalISel/GenericMachineInstrs.h:888]: Inconsistent capitalization with other ops, GShl?

File: llvm/lib/CodeGen/GlobalISel/CombinerHelperVectorOps.cpp [hunk 2]
```diff
@@ -400,3 +401,86 @@ bool CombinerHelper::matchInsertVectorElementOOB(MachineInstr &MI,
...
+bool CombinerHelper::matchAddOfVScale(const MachineOperand &MO,
+                                      BuildFnTy &MatchInfo) {
+  GAdd *Add = cast<GAdd>(MRI.getVRegDef(MO.getReg()));
+  GVScale *LHSVScale = cast<GVScale>(MRI.getVRegDef(Add->getLHSReg()));
+  GVScale *RHSVScale = cast<GVScale>(MRI.getVRegDef(Add->getRHSReg()));
+
+  Register Dst = Add->getReg(0);
+
+  if (!MRI.hasOneNonDBGUse(LHSVScale->getReg(0)) ||
+      !MRI.hasOneNonDBGUse(RHSVScale->getReg(0)))
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelperVectorOps.cpp:414]: Does the HasOneUse predicate added in d0dc29c2084a18c33b1b5b1cad9fd42215869746 need additional wiring to be used in combines?

Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelperVectorOps.cpp:414]: From a brief reading, it seems to be limited to instruction emitters and cannot be used in the combiner.

**Review by arsenm — APPROVED** [review]
Would be helped by a hasOneUse operator

**Pull request merged**